### PR TITLE
NAS-129123 / 24.10 / Expand error recovery in AD health checks

### PR DESCRIPTION
**Pull request by anodos325**
This primarily adds two new features to our AD health checks

* Check whether the secrets.tdb file exists and has a valid machine account password. If it's missing, then try to restore from backup.

* Check whether we have a stored kerberos keytab for the AD domain. If it's missing, reconstruct it from our machine account password in the secrets.tdb file.

This commit also refactors the kerberos plugin to move many methods into general-purpose krb5 utils that can be tested more easily in isolation.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/activedirectory_/health.py [hunk 1]
```diff
@@ -28,6 +29,88 @@ def winbind_status(self, check_trust=False):
...
+                'domain': ad_config['domainname'].upper(),
+                'kdc': dc
+            }}
+        })
+
+        try:
+            self.middleware.call_sync('kerberos.kdestroy', {'ccache': 'TEMP'})
+        except Exception:
+            self.logger.debug("Failed to destroy temporary ccache", exc_info=True)
+
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/activedirectory_/health.py:113]: flake8 issue at line 125: change to `if isinstance(out[key], list):`

File: src/middlewared/middlewared/plugins/directoryservices_/secrets.py [hunk 2]
```diff
@@ -113,6 +113,12 @@ async def get_ldap_idmap_secret(self, domain, user_dn):
         """
         return await self.__fetch(f'SECRETS/GENERIC/IDMAP_LDAP_{domain.upper()}/{user_dn}')
 
+    async def get_machine_secret(self, domain):
```
Comment (mgrimesix) [inline src/middlewared/middlewared/plugins/directoryservices_/secrets.py:116]: flake8 on line 8:  `TDBError` imported but not used.

File: src/middlewared/middlewared/utils/directoryservices/krb5.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+from tempfile import NamedTemporaryFile
+
+
+def __tmp_krb5_keytab() -> str:
+    """
+    Create a temporary keytab file with appropriate header
+    """
+    tmpfile = NamedTemporaryFile(delete=False)
+    try:
+        tmpfile.write(b'\x05\x02')
```
Comment (mgrimesix) [inline src/middlewared/middlewared/utils/directoryservices/krb5.py:25]: Why not do the write under a `with` context?

**Review by mgrimesix — COMMENTED** [review]
This was not a detailed review.  I _might_ circle back.

File: src/middlewared/middlewared/plugins/activedirectory.py [hunk 4]
```diff
@@ -782,19 +781,16 @@ async def __start(self, job):
...
-                    'commands issued outside of public APIs) while lacking a configured kerberos '
-                    'principal, which is required maintain a stable domain connection. Disabling service.'
-                )
+                if not await self.middleware.call('kerberos.keytab.query', [['AD_MACHINE_ACCOUNT']]):
+                    # Force writing of keytab based on stored secrets to our config file.
+                    await self.middleware.call('activedirectory.check_machine_account_keytab', ad['domainname'])
+
+                await self.middleware.call('datastore.update', self._config.datastore, ad['id'], {
+                    'kerberos_principal': machine_acct
+                }, {'prefix': 'ad_'})
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/activedirectory.py:794]: We still do, but via catch-all in `activedirectory.update`.

File: src/middlewared/middlewared/plugins/ldap.py [hunk 5]
```diff
@@ -733,21 +733,32 @@ def port_is_listening(self, host, port, timeout=1):
...
+                'bindpw': ldap_config.get('bindpw', ''),
+                'kerberos_realm': ldap_config.get('kerberos_realm', ''),
+                'kerberos_principal': ldap_config.get('kerberos_principal', ''),
+            }
+        }
+        cred = await self.middleware.call('kerberos.get_cred', payload)
+        await self.middleware.call('kerberos.do_kinit', {'krb5_cred': cred})
+
     @private
     async def validate_credentials(self, ldap_config=None):
```
Comment (bmeagherix) [inline src/middlewared/middlewared/plugins/ldap.py:758]: Comment?

File: src/middlewared/middlewared/plugins/directoryservices_/cache.py [hunk 6]
```diff
@@ -144,7 +142,6 @@ async def query(self, objtype, filters, options):
 
         `objtype`: 'USERS' or 'GROUPS'
         """
-        res = []
```
Comment (bmeagherix) [inline src/middlewared/middlewared/plugins/directoryservices_/cache.py:147]: Unused `res` is removed - fine.  However, it is referenced in a comment a few lines down.  Does this need to be updated?

File: src/middlewared/middlewared/utils/directoryservices/krb5.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    `filters` - query-filters
+    """
+    kt_list = ktutil_list_impl(keytab_file)
+    to_keep = filter_list(kt_list, filters)
+    to_remove = [entry['slot'] for entry in kt_list if entry not in to_keep]
+
+    if len(kt_list) == len(to_remove):
+        # Let caller know that keytab would be empty. If we were to follow
+        # through with this, caller would receive # keytab containing only
+        # `b'\x05\x02'`
```
Comment (bmeagherix) [inline src/middlewared/middlewared/utils/directoryservices/krb5.py:257]:
```
        # `b'\x05\x02'` (KRB5_KT_VNO)
```

**Review by bmeagherix — APPROVED** [review]
Approving because all comments are suggestions/cleanup, and not _essential_.

**Review by mgrimesix — APPROVED** [review]
Approve also since most my suggestions were flake8 fixes.

File: src/middlewared/middlewared/plugins/kerberos.py [hunk 8]
```diff
@@ -228,12 +123,26 @@ def write_realms(krb_file):
...
+    @accepts(
+        Dict(
+            'kerberos-options',
+            Str('ccache', enum=[x.name for x in krb5ccache], default=krb5ccache.SYSTEM.name),
+            Int('ccache_uid', default=0),
+            register=True,
+        ),
+        Bool('raise_error', default=True)
+    )
+    async def check_ticket(self, data, raise_error):
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/kerberos.py:133]: If we get too creative we can start to hit DNS timeouts. I mostly use this to have failure points if we expect to have a kerberos ticket, but it's not present.

File: src/middlewared/middlewared/plugins/kerberos.py [hunk 9]
```diff
@@ -1148,78 +892,55 @@ async def kerberos_principal_choices(self):
...
-            encoded_keytab = base64.b64encode(f.read())
+        if not os.path.exists(KRB_Keytab.SYSTEM.value):
+            self.logger.warning('System keytab is missing. Unable to extract AD machine account keytab.')
+            return
 
-        if not encoded_keytab:
-            self.logger.debug(f"Failed to generate b64encoded version of {keytab['SAMBA'].name}")
-            return False
+        ad = self.middleware.call_sync('activedirectory.config')
+        ad_kt_bytes = extract_from_keytab(KRB_Keytab.SYSTEM.value, [['principal', 'Crin', ad['netbiosname']]])
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/kerberos.py:925]: The `C` prefix is used to indicate case-insensitive. This is documented in developer documentation for `filter-list`.

[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/kerberos.py:925]: >  Shouldn't we include the trailing $ here?

No. There are a variety of different principals used when keytab is generated on domain join. Some use account name (netbiosname$), others use the hostname / netbiosname directly.

**Pull request merged**